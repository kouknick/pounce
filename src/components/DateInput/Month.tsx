import React from 'react';
import styled from '@emotion/styled';
import dayjs, { Dayjs } from 'dayjs';
import { chunk } from 'lodash';
import Flex from '../Flex';
import Day from './Day';
import { noop } from '../../utils/helpers';
export interface MonthProps {
  year: number;
  month: number;
  daySelected?: Dayjs;
  onDaySelect?: (date: Dayjs) => void;
}

const Abbr = styled.abbr`
  font-size: ${({ theme }) => theme.fontSizes.small};
  font-weight: ${({ theme }) => theme.fontWeights.bold};
  text-decoration: none;
`;

const Body = styled.div`
  display: table-row-group;
`;

const WeekRow = styled.div`
  display: table-row;
`;

const DAYS = [
  { abbr: 'Mo', name: 'Monday' },
  { abbr: 'Tu', name: 'Tuesday' },
  { abbr: 'We', name: 'Wednesday' },
  { abbr: 'Th', name: 'Thurdsday' },
  { abbr: 'Fr', name: 'Friday' },
  { abbr: 'Sa', name: 'Saturday' },
  { abbr: 'Su', name: 'Sunday' },
];

const getWeekCount = (year: number, month: number): number => {
  const firstDay = dayjs().month(month).year(year).date(1);
  const daysCount = firstDay.daysInMonth();
  const add = firstDay.day() === 0 ? 7 : firstDay.day();
  return Math.ceil((daysCount + add) / 7);
};

const Month: React.FC<MonthProps> = ({ year, month, daySelected, onDaySelect = noop }) => {
  const weeks = React.useMemo(() => {
    const weekCount = getWeekCount(year, month);
    const monthDate = dayjs().month(month).year(year).date(1);
    const daysCount = monthDate.daysInMonth();
    const start = monthDate.day();
    const days = Array(weekCount * 7);
    const step = start === 0 ? 7 : start;
    for (let index = 0; index <= daysCount; index++) {
      days[index + step - 2] = index;
    }
    return chunk(days, 7);
  }, [year, month]);

  return (
    <>
      <div>
        <Flex
          role="row"
          justify="space-around"
          mb={2}
          py={2}
          borderTop="1px solid"
          borderBottom="1px solid"
          borderColor="navyblue-300"
        >
          {DAYS.map(day => (
            <div role="columnheader" key={day.abbr}>
              <Abbr title={day.name}>{day.abbr}</Abbr>
            </div>
          ))}
        </Flex>
      </div>
      <Body>
        {weeks.map((week, monthIndex) => (
          // eslint-disable-next-line
          <WeekRow key={`${year}-${month}-${monthIndex}-week`}>
            {week.map((day, dayIndex) => (
              <Day
                onDaySelect={onDaySelect}
                daySelected={daySelected}
                year={year}
                month={month}
                day={day}
                key={
                  // eslint-disable-next-line
                  `${year}-${month}-${monthIndex}-${dayIndex}-week`
                }
              />
            ))}
          </WeekRow>
        ))}
      </Body>
    </>
  );
};

export default Month;
