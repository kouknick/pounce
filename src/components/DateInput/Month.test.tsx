import React from 'react';
import { renderWithTheme, fireEvent } from 'test-utils';
import dayjs from 'dayjs';
import Month from './Month';

const month = 10;
const year = 2020;
const day = 1;
const date = dayjs().date(day).month(month).year(year);

describe('Month', () => {
  it('renders', async () => {
    const { container } = await renderWithTheme(<Month month={month} year={year} />);
    expect(container).toMatchSnapshot();
  });

  it('renders with a selected day', async () => {
    const { container } = await renderWithTheme(
      <Month daySelected={date} month={month} year={year} />
    );
    expect(container).toMatchSnapshot();
  });

  it('allows selecting a date', async () => {
    const mock = jest.fn();
    const { findByLabelText } = await renderWithTheme(
      <Month daySelected={date} month={month} year={year} onDaySelect={mock} />
    );
    const cell = await findByLabelText('Su Nov 01 2020');
    await fireEvent.click(cell);
    expect(mock).toHaveBeenCalled();
  });
});
