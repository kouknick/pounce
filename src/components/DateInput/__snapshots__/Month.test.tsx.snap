// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month renders 1`] = `
.pounce-14 {
  margin-bottom: 8px;
  padding-top: 8px;
  padding-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-top: 1px solid;
  border-bottom: 1px solid;
  border-color: #3e516b;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
}

.pounce-0 {
  font-size: 0.75rem;
  font-weight: 700;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.pounce-112 {
  display: table-row-group;
}

.pounce-30 {
  -webkit-display: table-row;
  display: table-row;
}

.pounce-16 {
  display: table-cell;
  padding: 8px;
  vertical-align: middle;
  text-align: center;
  cursor: pointer;
  border-radius: 50%;
  cursor: default;
  pointer-events: none;
}

.pounce-16[aria-selected='true'],
.pounce-16[aria-selected='true']:hover {
  background-color: #0081c5;
}

.pounce-16:hover {
  background-color: #3e516b;
}

.pounce-28 {
  display: table-cell;
  padding: 8px;
  vertical-align: middle;
  text-align: center;
  cursor: pointer;
  border-radius: 50%;
}

.pounce-28[aria-selected='true'],
.pounce-28[aria-selected='true']:hover {
  background-color: #0081c5;
}

.pounce-28:hover {
  background-color: #3e516b;
}

<div>
  <div>
    <div
      class="pounce-14 pounce-15"
      role="row"
    >
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Monday"
        >
          Mo
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Tuesday"
        >
          Tu
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Wednesday"
        >
          We
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Thurdsday"
        >
          Th
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Friday"
        >
          Fr
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Saturday"
        >
          Sa
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Sunday"
        >
          Su
        </abbr>
      </div>
    </div>
  </div>
  <div
    class="pounce-112 pounce-113"
  >
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="false"
        aria-label="Su Nov 01 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        1
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 02 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        2
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 03 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        3
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 04 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        4
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 05 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        5
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 06 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        6
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 07 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        7
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 08 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        8
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 09 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        9
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 10 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        10
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 11 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        11
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 12 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        12
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 13 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        13
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 14 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        14
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 15 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        15
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 16 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        16
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 17 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        17
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 18 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        18
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 19 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        19
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 20 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        20
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 21 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        21
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 22 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        22
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 23 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        23
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 24 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        24
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 25 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        25
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 26 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        26
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 27 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        27
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 28 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        28
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 29 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        29
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 30 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        30
      </div>
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
    </div>
  </div>
</div>
`;

exports[`Month renders with a selected day 1`] = `
.pounce-14 {
  margin-bottom: 8px;
  padding-top: 8px;
  padding-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-top: 1px solid;
  border-bottom: 1px solid;
  border-color: #3e516b;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
}

.pounce-0 {
  font-size: 0.75rem;
  font-weight: 700;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.pounce-112 {
  display: table-row-group;
}

.pounce-30 {
  -webkit-display: table-row;
  display: table-row;
}

.pounce-16 {
  display: table-cell;
  padding: 8px;
  vertical-align: middle;
  text-align: center;
  cursor: pointer;
  border-radius: 50%;
  cursor: default;
  pointer-events: none;
}

.pounce-16[aria-selected='true'],
.pounce-16[aria-selected='true']:hover {
  background-color: #0081c5;
}

.pounce-16:hover {
  background-color: #3e516b;
}

.pounce-28 {
  display: table-cell;
  padding: 8px;
  vertical-align: middle;
  text-align: center;
  cursor: pointer;
  border-radius: 50%;
}

.pounce-28[aria-selected='true'],
.pounce-28[aria-selected='true']:hover {
  background-color: #0081c5;
}

.pounce-28:hover {
  background-color: #3e516b;
}

<div>
  <div>
    <div
      class="pounce-14 pounce-15"
      role="row"
    >
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Monday"
        >
          Mo
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Tuesday"
        >
          Tu
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Wednesday"
        >
          We
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Thurdsday"
        >
          Th
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Friday"
        >
          Fr
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Saturday"
        >
          Sa
        </abbr>
      </div>
      <div
        role="columnheader"
      >
        <abbr
          class="pounce-0 pounce-1"
          title="Sunday"
        >
          Su
        </abbr>
      </div>
    </div>
  </div>
  <div
    class="pounce-112 pounce-113"
  >
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="false"
        aria-label="Su Nov 01 2020"
        aria-selected="true"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        1
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 02 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        2
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 03 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        3
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 04 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        4
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 05 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        5
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 06 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        6
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 07 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        7
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 08 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        8
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 09 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        9
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 10 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        10
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 11 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        11
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 12 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        12
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 13 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        13
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 14 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        14
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 15 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        15
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 16 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        16
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 17 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        17
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 18 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        18
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 19 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        19
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 20 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        20
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 21 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        21
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 22 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        22
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 23 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        23
      </div>
      <div
        aria-disabled="false"
        aria-label="Tu Nov 24 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        24
      </div>
      <div
        aria-disabled="false"
        aria-label="We Nov 25 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        25
      </div>
      <div
        aria-disabled="false"
        aria-label="Th Nov 26 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        26
      </div>
      <div
        aria-disabled="false"
        aria-label="Fr Nov 27 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        27
      </div>
      <div
        aria-disabled="false"
        aria-label="Sa Nov 28 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        28
      </div>
      <div
        aria-disabled="false"
        aria-label="Su Nov 29 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        29
      </div>
    </div>
    <div
      class="pounce-30 pounce-31"
    >
      <div
        aria-disabled="false"
        aria-label="Mo Nov 30 2020"
        aria-selected="false"
        class="pounce-28 pounce-29"
        role="gridcell"
      >
        30
      </div>
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
      <div
        aria-disabled="true"
        class="pounce-16 pounce-17"
      />
    </div>
  </div>
</div>
`;
